File: src/main/java/io/klbz/curie/Pair.java
package io.klbz.curie;

import java.util.Objects;
import java.util.function.*;

public final class Pair<F, S> {
	private final F first;
	private final S second;

	private Pair(F first, S second){
		this.first = first;
		this.second = second;
	}

	public static <F, S> Pair<F, S> of(F first, S second){ return new Pair<>(first, second); }

	public static <F> PairOfAndBuilder<F> of(F first){ return new PairOfAndBuilder<>(first); }

	public static <F> PairBuilderMissingSecond<F> first(F first){ return new PairBuilderMissingSecond<>(first); }

	public static <S> PairBuilderMissingFirst<S> second(S second){ return new PairBuilderMissingFirst<>(second); }

	public <FPrime> Pair<FPrime, S> replaceFirst(FPrime newFirst){ return Pair.of(newFirst, second); }

	public <SPrime> Pair<F, SPrime> replaceSecond(SPrime newSecond){ return Pair.of(first, newSecond); }

	public FirstFocus<F, S> focusFirst(){ return FirstFocus.on(this); }

	public SecondFocus<F, S> focusSecond(){ return SecondFocus.on(this); }

	public F isolateFirst(){ return first; }

	public S isolateSecond(){ return second; }

	public <FPrime> Pair<FPrime, S> mapFirst(Function<? super F, FPrime> f){ return of(f.apply(first), second); }

	public <SPrime> Pair<F, SPrime> mapSecond(Function<? super S, SPrime> f){ return of(first, f.apply(second)); }

	public <T> T collapse(BiFunction<? super F, ? super S, T> f){ return f.apply(first, second); }

	public boolean satisfies(BiPredicate<? super F, ? super S> p){ return p.test(first, second); }

	public Pair<F, S> withFirstDo(Consumer<F> doF){
		doF.accept(first);
		return this;
	}

	public Pair<F, S> withSecondDo(Consumer<S> doF){
		doF.accept(second);
		return this;
	}

	public Pair<F, S> withBothDo(BiConsumer<F, S> doF){
		doF.accept(first, second);
		return this;
	}

	@Override
	public boolean equals(Object o){
		if(this == o) return true;
		if(o == null) return false;

		if(o.getClass() == FirstFocus.class){
			FirstFocus<?, ?> that = (FirstFocus<?, ?>) o;
			return Objects.equals(this, that.unfocus());
		}

		if(o.getClass() == SecondFocus.class){
			SecondFocus<?, ?> that = (SecondFocus<?, ?>) o;
			return Objects.equals(this, that.unfocus());
		}

		if(getClass() != o.getClass()) return false;

		Pair<?, ?> pair = (Pair<?, ?>) o;
		return Objects.equals(first, pair.first) &&
		       Objects.equals(second, pair.second);
	}

	@Override
	public int hashCode(){ return Objects.hash(first, second); }

	@Override
	public String toString(){ return "Pair{" + first + "," + second + '}'; }

	public static final class PairOfAndBuilder<F> {
		private final F first;

		private PairOfAndBuilder(F first){ this.first = first; }

		public <S> Pair<F, S> and(S second){ return Pair.of(first, second); }
	}

	public static final class PairBuilderMissingSecond<F> {
		private final F first;

		private PairBuilderMissingSecond(F first){ this.first = first; }

		public <S> Pair<F, S> second(S second){ return Pair.of(first, second); }
	}

	public static final class PairBuilderMissingFirst<S> {
		private final S second;

		private PairBuilderMissingFirst(S second){ this.second = second; }

		public <F> Pair<F, S> first(F first){ return Pair.of(first, second); }
	}
}

File: src/main/java/io/klbz/curie/Pair.kt
package io.klbz.curie

import java.util.*
import java.util.function.*
import java.util.function.Function

class Pair<F, S> private constructor(private val first: F, private val second: S) {

    fun <FPrime> replaceFirst(newFirst: FPrime) = of(newFirst, second)

    fun <SPrime> replaceSecond(newSecond: SPrime) = of(first, newSecond)

    fun focusFirst() = FirstFocus.on(this)

    fun focusSecond() = SecondFocus.on(this)

    fun isolateFirst() = first

    fun isolateSecond() = second

    fun <FPrime> mapFirst(f: Function<in F, FPrime>) = of(f.apply(first), second)

    fun <SPrime> mapSecond(f: Function<in S, SPrime>) = of(first, f.apply(second))

    fun <T> collapse(f: BiFunction<in F, in S, T>) = f.apply(first, second)

    fun satisfies(p: BiPredicate<in F, in S>): Boolean = p.test(first, second)

    fun withFirstDo(doF: Consumer<F>) = also { doF.accept(first) }

    fun withSecondDo(doF: Consumer<S>) = also { doF.accept(second) }

    fun withBothDo(doF: BiConsumer<F, S>) = also { doF.accept(first, second) }

    //region Kotlin-specific overloads

    @JvmSynthetic
    fun <FPrime> mapFirst(f: (F) -> FPrime) = of(f(first), second)

    @JvmSynthetic
    fun <SPrime> mapSecond(f: (S) -> SPrime) = of(first, f(second))

    @JvmSynthetic
    fun <T> collapse(f: (F, S) -> T) = f(first, second)

    @JvmSynthetic
    fun satisfies(p: (F, S) -> Boolean): Boolean = p(first, second)

    @JvmSynthetic
    fun withFirstDo(doF: (F) -> kotlin.Unit) = also { doF(first) }

    @JvmSynthetic
    fun withSecondDo(doF: (S) -> kotlin.Unit) = also { doF(second) }

    @JvmSynthetic
    fun withBothDo(doF: (F, S) -> kotlin.Unit) = also { doF(first, second) }

    //endregion

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false

        if (other.javaClass == FirstFocus::class.java) {
            val that = other as FirstFocus<*, *>?
            return this == that!!.unfocus()
        }

        if (other.javaClass == SecondFocus::class.java) {
            val that = other as SecondFocus<*, *>?
            return this == that!!.unfocus()
        }

        if (javaClass != other.javaClass) return false

        val pair = other as Pair<*, *>?
        return first == pair!!.first && second == pair.second
    }

    override fun hashCode(): Int = Objects.hash(first, second)

    override fun toString() = "Pair{$first,$second}"

    class PairOfAndBuilder<F> internal constructor(private val first: F) {
        fun <S> and(second: S) = of(first, second)
    }

    class PairBuilderMissingSecond<F> internal constructor(private val first: F) {
        fun <S> second(second: S) = of(first, second)
    }

    class PairBuilderMissingFirst<S> internal constructor(private val second: S) {
        fun <F> first(first: F) = of(first, second)
    }

    companion object {
        @JvmStatic
        fun <F, S> of(first: F, second: S) = Pair(first, second)

        @JvmStatic
        fun <F> of(first: F) = PairOfAndBuilder(first)

        @JvmStatic
        fun <F> first(first: F) = PairBuilderMissingSecond(first)

        @JvmStatic
        fun <S> second(second: S) = PairBuilderMissingFirst(second)
    }
}
